### PR TITLE
Added logging for view compilation

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue - #3755 

@pranavkm @rynowak

### CODE REVIEW

**Comment by davidfowl** [comment]
Finally!

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/RazorCompilationService.cs [hunk 1]
```diff
@@ -49,7 +54,11 @@ public CompilationResult Compile(RelativeFileInfo file)
             GeneratorResults results;
             using (var inputStream = file.FileInfo.CreateReadStream())
             {
+                _logger.RazorFileToCodeCompilationStart(file.RelativePath);
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/RazorCompilationService.cs]: Not good enough, needs the time taken to compile.

**Comment by davidfowl** [comment]
These just need timings and it looks perfect :smile:

File: test/WebSites/RazorPageExecutionInstrumentationWebSite/project.json [hunk 2]
```diff
@@ -10,7 +10,8 @@
     "Microsoft.AspNet.Mvc.TestConfiguration": "1.0.0",
     "Microsoft.AspNet.Server.Kestrel": "1.0.0-*",
     "Microsoft.AspNet.StaticFiles": "1.0.0-*",
-    "Microsoft.Extensions.DiagnosticAdapter": "1.0.0-*"
+    "Microsoft.Extensions.DiagnosticAdapter": "1.0.0-*",
+    "Microsoft.Extensions.Logging.Testing": "1.0.0-rc2-*"
```
Comment (pranavkm) [inline test/WebSites/RazorPageExecutionInstrumentationWebSite/project.json]: `1.0.0-*`

File: test/WebSites/RazorPageExecutionInstrumentationWebSite/TestRazorCompilationService.cs [hunk 3]
```diff
@@ -15,7 +16,7 @@ public class TestRazorCompilationService : RazorCompilationService
             ICompilationService compilationService,
             IMvcRazorHost razorHost,
             IRazorViewEngineFileProviderAccessor fileProviderAccessor)
-            : base(compilationService, razorHost, fileProviderAccessor)
+            : base(compilationService, razorHost, fileProviderAccessor, NullLoggerFactory.Instance)
```
Comment (pranavkm) [inline test/WebSites/RazorPageExecutionInstrumentationWebSite/TestRazorCompilationService.cs]: DI should inject an `ILoggerFactory`. You shouldn't need the NullLoggerFactory.

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs [hunk 4]
```diff
@@ -89,6 +94,8 @@ public CompilationResult Compile(RelativeFileInfo fileInfo, string compilationCo
                 throw new ArgumentNullException(nameof(compilationContent));
             }
 
+            _logger.GeneratedCodeToAssemblyCompilationStart(fileInfo.RelativePath);
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs]: Look at https://github.com/aspnet/Mvc/commit/c125b4e59badf551c85bbc8d6ae8cb133fe6de7c for the timing pattern.

**Comment by pranavkm** [comment]
:watch:

**Comment by ajaybhargavb** [comment]
Updated.

File: src/Microsoft.AspNet.Mvc.Razor/Logging/RazorCompilationServiceLoggerExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        private static readonly double TimestampToTicks = Stopwatch.Frequency / 10000000.0;
+        private static readonly Action<ILogger, string, Exception> _razorFileToCodeCompilationStart;
+        private static readonly Action<ILogger, string, double, Exception> _razorFileToCodeCompilationEnd;
+
+        static RazorCompilationServiceLoggerExtensions()
+        {
+            _razorFileToCodeCompilationStart = LoggerMessage.Define<string>(
+                LogLevel.Debug,
+                1,
+                "Compilation of the razor file at '{FilePath}' started.");
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Logging/RazorCompilationServiceLoggerExtensions.cs]: `Razor` (casing)

File: src/Microsoft.AspNet.Mvc.Razor/Logging/RazorViewEngineLoggerExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNet.Mvc.Logging
+{
+    public static class RazorViewEngineLoggerExtensions
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Logging/RazorViewEngineLoggerExtensions.cs]: `CompilerCacheExtensions`

File: src/Microsoft.AspNet.Mvc.Razor/Logging/RazorCompilationServiceLoggerExtensions.cs (hunk 5 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Logging/RazorCompilationServiceLoggerExtensions.cs]: > Parsing Razor file at `{FilePath}' started.

File: src/Microsoft.AspNet.Mvc.Razor/Logging/RazorCompilationServiceLoggerExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+            _razorFileToCodeCompilationEnd = LoggerMessage.Define<string, double>(
+                LogLevel.Debug,
+                2,
+                "Compilation of the razor file at '{FilePath}' completed in {ElapsedMilliseconds}ms.");
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Logging/RazorCompilationServiceLoggerExtensions.cs]: > Parsing Razor file at '{FilePath}' completed in..

File: src/Microsoft.AspNet.Mvc.Razor/Logging/RoslynCompilationServiceLoggerExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        private static readonly double TimestampToTicks = Stopwatch.Frequency / 10000000.0;
+        private static readonly Action<ILogger, string, Exception> _generatedCodeToAssemblyCompilationStart;
+        private static readonly Action<ILogger, string, double, Exception> _generatedCodeToAssemblyCompilationEnd;
+
+        static RoslynCompilationServiceLoggerExtensions()
+        {
+            _generatedCodeToAssemblyCompilationStart = LoggerMessage.Define<string>(
+                LogLevel.Debug,
+                1,
+                "Compilation of the generated code to assembly for the razor file at '{FilePath}' started.");
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Logging/RoslynCompilationServiceLoggerExtensions.cs]: > Compilation of the generated code for Razor file at '{FilePath}`

**Comment by pranavkm** [comment]
:watch: for the log messages.

File: src/Microsoft.AspNet.Mvc.Razor/Logging/RazorViewEngineLoggerExtensions.cs (hunk 6 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Razor/Logging/RazorViewEngineLoggerExtensions.cs]: I thought we follow the pattern of using the name of the file where logging happens.

**Comment by ajaybhargavb** [comment]
Updated.

File: src/Microsoft.AspNet.Mvc.Razor/Logging/RazorViewEngineLoggerExtensions.cs (hunk 6 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Logging/RazorViewEngineLoggerExtensions.cs]: Oops, completely missed this bit. We have two levels of logging and I think what you have is the right place to log.

**Comment by pranavkm** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.Razor/Logging/RoslynCompilationServiceLoggerExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        private static readonly double TimestampToTicks = Stopwatch.Frequency / 10000000.0;
+        private static readonly Action<ILogger, string, Exception> _generatedCodeToAssemblyCompilationStart;
+        private static readonly Action<ILogger, string, double, Exception> _generatedCodeToAssemblyCompilationEnd;
+
+        static RoslynCompilationServiceLoggerExtensions()
+        {
+            _generatedCodeToAssemblyCompilationStart = LoggerMessage.Define<string>(
+                LogLevel.Debug,
+                1,
+                "Compilation of the generated code for the Razor file at '{FilePath}' started.");
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Logging/RoslynCompilationServiceLoggerExtensions.cs]: Do we refer to "Razor" when logging elsewhere? The alternative would be ".cshtml"; that might be more easily understood.

Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Logging/RoslynCompilationServiceLoggerExtensions.cs]: We don't have Razor specific logging elsewhere, so no. However we do use that term in our docs
For .e.g
> render a view that uses Razor syntax

**Comment by davidfowl** [comment]
@ajaybhargavb Can you paste a sample of console output?

**Comment by ajaybhargavb** [comment]
@davidfowl 

image

**Pull request merged**